internal/vault/client.go

```diff
@@ -173,6 +174,37 @@ type defaultClient struct {
 	mu                 sync.RWMutex
 }
 
+// Validate the client, returning an error for any validation failures.
+// Typically, an invalid Client would be discarded and replaced with a new
+// instance.
+func (c *defaultClient) Validate() error {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	if c.authSecret == nil {
+		return fmt.Errorf("auth secret not set, never logged in")
+	}
+
+	if !c.skipRenewal {
+		if c.lastWatcherErr != nil {
+			return c.lastWatcherErr
+		}
+		if c.watcher == nil {
+			return errors.New("lifetime watcher not set")
+		}
+	}
+
+	if expired, err := c.checkExpiry(0); expired || err != nil {
+		var errs error
+		if expired {
+			errs = errors.Join(errs, errors.New("client token expired"))
+		}
+		return errors.Join(errs, err)
```

Comment:
This feels a little clunky to read, unless I'm misreading `checkExpiry()` it will never return `expired=true && err != nil`
```suggestion
		if expired {
			return errors.New("client token expired"))
		}
		return err
```

Reply:
Fair point, but I kind of like the idea of getting the additional error that may have been encountered when checking the expiration. It could provide a bit more context.